cloud/store/client.go

```diff
@@ -30,63 +35,114 @@ func NewClient(localService Cluster) *Client {
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer conn.Close()
 	c := cmd.NewClusterServiceClient(conn)
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	r, err := c.JoinPeer(ctx, req)
+	return c.NotifyPeer(ctx, req)
+}
+
+func (cl *Client) join(ctx context.Context, leaderAddr string, req *cmd.JoinPeerRequest) (*cmd.JoinPeerResponse, error) {
+	log.Printf("client join: %s %+v\n", leaderAddr, req)
+	addr, err := cl.rpcAddressFromRAFT(leaderAddr)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	if addr == cl.localService.address {
+		return cl.localService.JoinPeer(ctx, req)
 	}
-	log.Printf("join response: %s", r.String())
-	return nil
+
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	c := cmd.NewClusterServiceClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	return c.JoinPeer(ctx, req)
 }
 
-func (cl Client) Remove(leaderAddress string, req *cmd.RemovePeerRequest) error {
+func (cl *Client) remove(ctx context.Context, leaderAddress string, req *cmd.RemovePeerRequest) (*cmd.RemovePeerResponse, error) {
 	log.Printf("client remove: %s %+v\n", leaderAddress, req)
-	if cl.localService.address == leaderAddress {
-		return cl.localService.Remove(req.Id)
+	addr, err := cl.rpcAddressFromRAFT(leaderAddress)
+	if err != nil {
+		return nil, err
+	}
+	if cl.localService.address == addr {
+		return cl.localService.RemovePeer(ctx, req)
 	}
-	// lAddr := cl.leaderAddress()
-	// if lAddr == "" {
-	// 	return err
-	// }
-	conn, err := grpc.Dial(leaderAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer conn.Close()
 	c := cmd.NewClusterServiceClient(conn)
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	r, err := c.RemovePeer(ctx, req)
-	if err != nil {
-		return err
-	}
-	log.Printf("join response: %s", r.String())
-	return nil
+	return c.RemovePeer(ctx, req)
 }
 
-// TODO-RAFT
-// leaderAddress is just a workaround to get address of leader rpc  service
-func (cl Client) leaderAddress() string {
-	lAddr, _ := cl.localService.Raft.LeaderWithID()
-	if lAddr == "" {
-		return ""
+// rpcAddressFromRAFT returns the RPC address based on the provided RAFT address.
+// In a real cluster, the RPC port is the same for all nodes.
+// In a local environment, the RAFT ports need to be different. Specifically,
+// we calculate the RPC port as the RAFT port + 1.
+func (cl *Client) rpcAddressFromRAFT(raftAddr string) (string, error) {
+	host, port, err := net.SplitHostPort(raftAddr)
+	if err != nil {
+		return "", err
 	}
-	xs := strings.Split(string(lAddr), ":")
-	return xs[0] + ":2" + xs[1][1:]
+	if !cl.localService.isLocal {
```

Comment:
I don't understand why we need this particular `isLocal` and `isLocalhost` in other functions. By default we use default port and it's otherwise specified on the command line.

I see the possible limitations
1. Someone defined a different loopback address that 127.0.0.1 and our current logic will not detect it (nor should it)
2. Someone is using command line parameter to specify `127.0.0.1:12345` and we will override it with `localService.rpcPort` from my understanding of the code.


Also the claim
> // In a real cluster, the RPC port is the same for all nodes.
I think is wrong. In the default configuration yes, but if we give our users the ability to change it then we should be able to handle it.

Reply:
The 'isLocal' flag is used to explicitly indicate a local host environment, particularly when running a cluster on your local machine using command-line tools. This flag simplifies local testing of the application, eliminating the need for each node keep track of each individual RAFT and RPC ports.

> I think is wrong. In the default configuration yes, but if we give our users the ability to change it then we should be able to handle it.

In an actual cluster, all nodes share the same RPC port for simplicity and consistency. The use of different ports through flags is solely for local machine testing purposes. By maintaining uniform port configurations, we simplify both the cluster's configuration and its overall management. Each node only needs to be aware of the address, as it can autonomously determine the appropriate port, similar to how it's handled in the case of HTTP